sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,315 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package arm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+)
+
+const (
+	providersKey             = "providers"
+	subscriptionsKey         = "subscriptions"
+	resourceGroupsLowerKey   = "resourcegroups"
+	locationsKey             = "locations"
+	builtInResourceNamespace = "Microsoft.Resources"
+)
+
+var (
+	RootResourceIdentifier = &ResourceIdentifier{
+		parent:       nil,
+		resourceType: TenantResourceType,
+		name:         "",
+	}
+
+	SubscriptionResourceType  = NewResourceType("Microsoft.Resources", "subscriptions")
+	ResourceGroupResourceType = NewResourceType("Microsoft.Resources", "resourceGroups")
+	TenantResourceType        = NewResourceType("Microsoft.Resources", "tenants")
+	ProvidersResourceType     = NewResourceType("Microsoft.Resources", "providers")
+)
+
+// ResourceIdentifier represents a resource ID
+type ResourceIdentifier struct {
+	parent            *ResourceIdentifier
+	subscriptionId    string
+	provider          string
+	resourceGroupName string
+	location          string
+	resourceType      ResourceType
+	name              string
+	isChild           bool
+
+	stringValue *string
+}
+
+func (id ResourceIdentifier) Parent() *ResourceIdentifier {
+	return id.parent
+}
+
+func (id ResourceIdentifier) SubscriptionId() string {
+	return id.subscriptionId
+}
+
+func (id ResourceIdentifier) Provider() string {
+	return id.provider
+}
+
+func (id ResourceIdentifier) ResourceGroupName() string {
+	return id.resourceGroupName
+}
+
+func (id ResourceIdentifier) Location() string {
+	return id.location
+}
+
+func (id ResourceIdentifier) ResourceType() ResourceType {
+	return id.resourceType
+}
+
+func (id ResourceIdentifier) Name() string {
+	return id.name
+}
+
+func newResourceIdentifier(parent *ResourceIdentifier, resourceTypeName string, resourceName string) *ResourceIdentifier {
+	id := &ResourceIdentifier{}
+	id.init(parent, chooseResourceType(resourceTypeName, parent), resourceName, true)
+	return id
+}
+
+func newResourceIdentifierWithResourceType(parent *ResourceIdentifier, resourceType ResourceType, resourceName string) *ResourceIdentifier {
+	id := &ResourceIdentifier{}
+	id.init(parent, resourceType, resourceName, true)
+	return id
+}
+
+func newResourceIdentifierWithProvider(parent *ResourceIdentifier, providerNamespace, resourceTypeName, resourceName string) *ResourceIdentifier {
+	id := &ResourceIdentifier{}
+	id.init(parent, NewResourceType(providerNamespace, resourceTypeName), resourceName, false)
+	return id
+}
+
+func chooseResourceType(resourceTypeName string, parent *ResourceIdentifier) ResourceType {
+	switch strings.ToLower(resourceTypeName) {
```

Comment:
You want [`strings.EqualFold`](https://pkg.go.dev/strings#EqualFold) instead of all these calls to `strings.ToLower`